qiime/sdk/result.py

```diff
@@ -0,0 +1,212 @@
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import collections
+import distutils.dir_util
+import functools
+import os
+import uuid
+
+import qiime.sdk
+import qiime.core.archiver
+import qiime.core.type
+
+# Note: Result, Artifact, and Visualization classes are in this file to avoid
+# circular dependencies between Result and its subclasses. Result is tightly
+# coupled to Artifact and Visualization because it is a base class and a
+# factory, so having the classes in the same file helps make this coupling
+# explicit.
+
+
+ResultMetadata = collections.namedtuple('ResultMetadata',
+                                        ['uuid', 'type', 'provenance'])
+
+
+class Result:
+    """Base class for QIIME 2 result classes (Artifact and Visualization).
+
+    This class is not intended to be instantiated. Instead, it acts as a public
+    factory and namespace for interacting with Artifacts and Visualizations in
+    a generic way. It also acts as a base class for code reuse and provides an
+    API shared by Artifact and Visualization.
+
+    """
+    @classmethod
+    def _is_valid_type(cls, type_):
+        """Subclasses should override this method."""
+        return True
+
+    @classmethod
+    def peek(cls, filepath):
+        uuid, type, provenance = qiime.core.archiver.Archiver.peek(filepath)
+        if not cls._is_valid_type(type):
+            raise TypeError(
+                "Cannot peek at filepath %r because %s does not support type "
+                "%r." % (filepath, cls.__name__, type))
+        return ResultMetadata(uuid=uuid, type=type, provenance=provenance)
+
+    @classmethod
+    def load(cls, filepath):
+        """Factory for loading Artifacts and Visualizations."""
+        archiver = qiime.core.archiver.Archiver.load(filepath)
+
+        if Artifact._is_valid_type(archiver.type):
+            result = Artifact.__new__(Artifact)
+        elif Visualization._is_valid_type(archiver.type):
+            result = Visualization.__new__(Visualization)
+        else:
+            raise TypeError(
+                "Cannot load filepath %r into an Artifact or Visualization "
+                "because type %r is not supported."
+                % (filepath, archiver.type))
+
+        if type(result) is not cls and cls is not Result:
+            raise TypeError(
+                "Attempting to load %s with `%s.load`. Use `%s.load` instead."
+                % (type(result).__name__, cls.__name__,
+                   type(result).__name__))
+
+        result._archiver = archiver
+        return result
+
+    @classmethod
+    def extract(cls, filepath, output_dir):
+        # `peek` first to ensure the type is valid for the `cls` this was
+        # called on.
+        cls.peek(filepath)
```

Comment:
The error message from `peek` will look a little weird here, but we can refactor that later.